lnwallet/channel_test.go

```diff
@@ -842,9 +851,217 @@ func TestCheckHTLCNumberConstraint(t *testing.T) {
+
+		return &lnwire.UpdateAddHTLC{
+			PaymentHash: paymentHash,
+			Amount:      amount,
+			Expiry:      uint32(5),
+		}, returnPreimage
+	}
+
+	// Create a test channel which will be used for the duration of this
+	// unittest. The channel will be funded evenly with Alice having 5 BTC,
+	// and Bob having 5 BTC. Set the respective channel reserves.
+	aliceChannel, bobChannel, cleanUp, err := createTestChannels(3, true)
+	if err != nil {
+		t.Fatalf("unable to create test channels: %v", err)
+	}
+	defer cleanUp()
+	aliceChannelReserve := btcutil.Amount(1000)
+	bobChannelReserve := btcutil.Amount(2000)
+	aliceChannel.channelState.OurChannelReserve = aliceChannelReserve
+	aliceChannel.channelState.TheirChannelReserve = bobChannelReserve
+	bobChannel.channelState.OurChannelReserve = bobChannelReserve
+	bobChannel.channelState.TheirChannelReserve = aliceChannelReserve
+
+	aliceAmount := aliceChannel.channelState.OurBalance
+	bobAmount := bobChannel.channelState.OurBalance
+
+	// Alice creates an HTLC which leaves her with the minimum channel
+	// reserve allowed.
+	htlc, preimage := createHTLC(0, aliceAmount-aliceChannelReserve)
+	if _, err := aliceChannel.AddHTLC(htlc); err != nil {
+		t.Fatalf("alice unable to add htlc: %v", err)
+	}
+	if _, err := bobChannel.ReceiveHTLC(htlc); err != nil {
+		t.Fatalf("bob unable to receive htlc: %v", err)
+	}
+	if err := forceStateTransition(aliceChannel, bobChannel); err != nil {
+		t.Fatalf("Can't update the channel state: %v", err)
+	}
+
+	// Settle HTLC and sign new commitment.
+	settleIndex, err := bobChannel.SettleHTLC(preimage)
+	if err != nil {
+		t.Fatalf("bob unable to settle inbound htlc: %v", err)
+	}
+	err = aliceChannel.ReceiveHTLCSettle(preimage, settleIndex)
+	if err != nil {
+		t.Fatalf("alice unable to accept settle of outbound htlc: %v", err)
+	}
+	if err := forceStateTransition(bobChannel, aliceChannel); err != nil {
+		t.Fatalf("Can't update the channel state: %v", err)
+	}
+
+	// Alice tries to add a 1 satoshi HTLC which should not be possible,
+	// as she would go below her channel reserve limit.
+	htlc, _ = createHTLC(0, 1)
+	// Since the check for the local reserve limit is done elsewhere,
+	// alice's call to AddHTLC would pass, therefore we skip that call.
+	// Bob's call to ReceiveHTLC should stop Alice's attempt though
+	if _, err := bobChannel.ReceiveHTLC(htlc); err == nil {
+		t.Fatalf("bob able to receive htlc when he should not")
+	}
+
+	aliceAmount = aliceChannel.channelState.OurBalance
+	bobAmount = bobChannel.channelState.OurBalance
+
+	// Bob creates an HTLC which leaves him with the minimum channel
+	// reserve allowed.
+	htlc, preimage = createHTLC(0, bobAmount-bobChannelReserve)
+	if _, err := bobChannel.AddHTLC(htlc); err != nil {
+		t.Fatalf("bob unable to add htlc: %v", err)
+	}
+	if _, err := aliceChannel.ReceiveHTLC(htlc); err != nil {
+		t.Fatalf("alice unable to receive htlc: %v", err)
+	}
+	if err := forceStateTransition(aliceChannel, bobChannel); err != nil {
```

Comment:
Due to the new state machine, the order of the arguments here will need to be swapped.